### PR TITLE
fix URL(use_appname=False) in shell

### LINKED ISSUE #419: URL(use_appname=False) bug
```
[pax@localhost py4web]$ ./py4web.py shell apps
Python 3.8.6 (default, Sep 25 2020, 00:00:00) 
[GCC 10.2.1 20200723 (Red Hat 10.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from py4web.core import URL
>>> URL(use_appname=False)
Traceback (most recent call last):
  File "/home/pax/.local/lib/python3.8/site-packages/bottle.py", line 1397, in __getattr__
    var = self.environ['bottle.request.ext.%s'%name]
KeyError: 'bottle.request.ext.app_name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/home/pax/Job/py4web/github/py4web/py4web/core.py", line 614, in URL
    if (request.app_name != "_default" and use_appname)
  File "/home/pax/.local/lib/python3.8/site-packages/bottle.py", line 1400, in __getattr__
    raise AttributeError('Attribute %r not defined.' % name)
AttributeError: Attribute 'app_name' not defined.
```

### PR DESCRIPTION
**Pull request by paolopas**
fix #419

### CODE REVIEW

**Comment by jpsteil** [comment]
I began getting this error this afternoon even when use_appname = True. I tried this fix, but ended up with this on the same line of code to fix:

if (hasattr(request, 'app_name') and request.app_name != "_default" and use_appname)

instead.

It's weird because all was working fine and out of the blue, this started happening.

**Comment by jpsteil** [comment]
Retract that, now my URLs aren't building properly.  Will post more when I get it working.

**Comment by jpsteil** [comment]
Looks like my issue is related to building URLs in a method called by Inject, and not related to this issue.  I will research my issue more and submit an issue if I can't resolve it.

**Comment by valq7711** [comment]
`request.app_name`  is set in action decorator, i.e. URL is intended to use in action

**Comment by paolopas** [comment]
@jpsteil my fix is simpler, request.app_name is never evaluated when use_appname == False,
take a look at diff

**Comment by jpsteil** [comment]
Paolo

You are correct. My errors had to do with my use of URL in an inject and
not use_appname. Please disregard my comments.

**Comment by paolopas** [comment]
@valq7711:
The fact that URL should be used into actions only might be questionable.
Surely the documentation does not make any mention of that and we know that "poor documentation makes code poor".
I think that fixing documentation is a good way to contribute to open source, (developers are lazy guys...), and in fact I found the problem testing some examples of docs/chapter-10.rst.

@mdipierro:
It is docs/chapter-10.rst the right place to state that? Because it is a big change from web2py.

Or maybe we can accomodate to let URL be used by the cli shell too?

Anyway you can change the description of this pull request to whatever you want and the fix still remain valid, it is a matter of efficiency on expression evaluation.

**Comment by paolopas** [comment]
@valq7711 please can you tell which one of the last two commit is the right one?

**Comment by paolopas** [comment]
I think it is time for a break...

**Comment by valq7711** [comment]
@paolopas 
The logic is as follows:
If use_appname is `None`, then it is not passed, in which case it should be set to `True` (default), unless the application is mapped to a domain (`x-py4web-appname` indicates that app is mapped).
So you're right, this:
https://github.com/web2py/py4web/blob/ad764fce99740d35293814221966c1ca7843a82e/py4web/core.py#L602-L604
may be collapsed to this:
```python
if use_appname is None: 
     use_appname = not request.headers.get("x-py4web-appname") 
```

**Comment by paolopas** [comment]
Someone said **Readabilty counts**

**Comment by paolopas** [comment]
Someone said **Readabilty counts**

File: py4web/core.py [hunk 1]
```diff
@@ -600,8 +600,7 @@ def URL(
     URL('a','b',vars=dict(x=1),use_appname=False) -> /{script_name?}/a/b?x=1
     """
     if use_appname is None:
-        use_appname = request.headers.get("x-py4web-appname")
-        use_appname = True if use_appname is None else not use_appname
+        use_appname = request.headers.get("x-py4web-appname", "") == ""
```
Comment (valq7711) [inline py4web/core.py:603]: Fine!

**Comment by valq7711** [comment]
@paolopas have you seen https://github.com/paolopas/py4web/pull/1/files?

File: py4web/core.py (hunk 1 above)
Comment (paolopas) [reply py4web/core.py:603]: Your version sounds good
use_appname = not request.headers.get("x-py4web-appname")

**Comment by jpsteil** [comment]
So, will this fix allow you to use URL in Inject?

**Comment by valq7711** [comment]
@jpsteil  Jim, show your snippet please, as I can't figure out the reason why it doesn't work

**Comment by valq7711** [comment]
@jpsteil  maybe you do something like `foo = URL()` in inject?

**Comment by jpsteil** [comment]
Val - I'll pull in the PR and give it a test later today.

-Jim

**Comment by paolopas** [comment]
has_appname added for clarity

**Comment by paolopas** [comment]
I hope that is not necessary to explain that in Python `A and B`
does not necessarily evaluates B (as for `A or B`)

**Comment by paolopas** [comment]
@valq7711 for me there is nothing more to do, do you still have some doubts?

**Comment by valq7711** [comment]
@paolopas  perfection is unattainable ;)
https://github.com/paolopas/py4web/pull/2/files

**Comment by mdipierro** [comment]
I think URL can be used in actions and functions/methods called by actions. This should be stated where URL is first used in the docs and in chapter 10. If URL is used in a place where app_name is not passed and undefined it should raise an error (with assert).

**Comment by paolopas** [comment]
A good place could be the function docstring too

**Comment by paolopas** [comment]
@mdipierro
I think there are many reason for URL to do not enforce app_name (as it does now):
1. be tolerant if there is no strict reason to be not make code more robust,
1. actually when we use URL in _default app, it behave as if no app_name
   is present, so why do not do the same in cases app_name
   is really missing? (for whatever reason)
1. maybe an action could be called in py4web shell environment, without
   the need to give py4web shell a virtual app_name (e.g. py4web-shell)

Otherwise the right place to document should be chapter 15 ("From web2py to py4web")
since enforcing app_name is a big change with respect to web2py.

**Pull request merged**